### PR TITLE
Feature/cart list

### PR DESCRIPTION
**Pull request by barshathakuri**
Add cart page

### CODE REVIEW

File: app/Base/components/Navbar/index.tsx [hunk 1]
```diff
@@ -134,6 +136,13 @@ function Navbar(props: Props) {
...
                         Login
                     </SmartButtonLikeLink>
+                    <Link to="/cart-page">
+                        <QuickActionButton
+                            name="cart"
+                        >
+                            <FaShoppingCart />
+                        </QuickActionButton>
+                    </Link>
                     {authenticated && user && (
```
Comment (AdityaKhatri) [inline app/Base/components/Navbar/index.tsx]: Use ButtonLikeLink

File: app/views/CartPage/index.tsx [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+import {
+    gql,
+    useMutation,
+    useQuery,
+} from '@apollo/client';
+import { AiTwotoneDelete } from 'react-icons/ai';
+import { FaHeart } from 'react-icons/fa';
+
+import styles from './styles.css';
+import { CartListQuery, CartListQueryVariables, DeleteCartItemsMutation, DeleteCartItemsMutationVariables } from '#generated/types';
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx]:
```suggestion
import {
    CartListQuery,
    CartListQueryVariables,
    DeleteCartItemsMutation,
    DeleteCartItemsMutationVariables,
} from '#generated/types';
```

File: app/views/CartPage/index.tsx [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+interface Book {
+    id: number;
+    price: number;
+    authors: Author[];
+    quantity: number;
+    image: Image;
+    title: string;
+    cartId: string;
+    removeBookFromCartList: (id: string) => void;
+}
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx:70]: These can be taken from generated which will reduce redundancy

File: app/views/CartPage/index.tsx [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+
+function CartContent(props: Book) {
+    const { id, price, authors, image, title, quantity, removeBookFromCartList, cartId } = props;
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx:73]: drop in multiple lines

File: app/views/CartPage/index.tsx [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    const authorsDisplay = React.useMemo(() => (
+        authors?.map((d) => d.name).join(', ')
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx:75]: make a separate utility function to join strings

File: app/views/CartPage/index.tsx [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+    ), [authors]);
+
+    const deleteCartItem = useCallback(() => {
+        removeBookFromCartList(cartId);
+    }, [cartId]);
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx:80]: set name of button to cartId and directly set removeBookFromCartList in onClick.

File: app/views/CartPage/index.tsx [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+                {image?.url ? (
+                    <img
+                        className={styles.image}
+                        src={image.url}
+                        alt={title}
+                    />
+                ) : (
+                    <div className={styles.noPreview}>
+                        Preview not available
+                    </div>
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx:94]: Use message component

File: app/views/CartPage/index.tsx [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+    const { data, refetch, loading } = useQuery<
+        CartListQuery,
+        CartListQueryVariables
+    >(CART_LIST, {
+        variables: { page, pageSize, email },
+    });
+
+    const removeCartItem = (id: string) => {
+        deleteCartItem({ variables: { id } });
+        refetch();
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx:164]: refetch should be done after delete is completed.

File: app/views/CartPage/index.tsx [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+                                        id={b.book.id}
+                                        title={b.book.title}
+                                        image={b.book.image}
+                                        price={b.book.price}
+                                        quantity={b.quantity}
+                                        authors={b.book.authors}
+                                        removeBookFromCartList={removeCartItem}
+                                        cartId={b.id}
+                                    />
+                                ))
```
Comment (AdityaKhatri) [inline app/views/CartPage/index.tsx:184]: define type of b and use ListView

**Review by AdityaKhatri — CHANGES_REQUESTED** [review]

File: app/Base/components/Navbar/index.tsx [hunk 10]
```diff
@@ -134,6 +136,13 @@ function Navbar(props: Props) {
                     >
                         Login
                     </SmartButtonLikeLink>
+                    <Link to="/cart-page">
+                        <QuickActionButton
+                            name="cart"
+                        >
+                            <FaShoppingCart />
+                        </QuickActionButton>
+                    </Link>
```
Comment (tnagorra) [inline app/Base/components/Navbar/index.tsx]: We should never nest button inside link.
The semantics is not correct.

File: app/Base/configs/routes.tsx [hunk 11]
```diff
@@ -58,13 +58,23 @@ const bookDetail = wrap({
     },
     visibility: 'is-anything',
 });
+const cartPage = wrap({
+    path: '/cart-page/',
```
Comment (tnagorra) [inline app/Base/configs/routes.tsx]: Let's remove `-page` from page url

File: app/Base/configs/routes.tsx [hunk 12]
```diff
@@ -58,13 +58,23 @@ const bookDetail = wrap({
...
+    title: 'My Cart Page',
```
Comment (tnagorra) [inline app/Base/configs/routes.tsx]: Let's remove `Page` from title

File: app/views/CartPage/index.tsx [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+import { Button, Container, NumberInput, TextOutput } from '@the-deep/deep-ui';
+import {
+    gql,
+    useMutation,
+    useQuery,
+} from '@apollo/client';
+import { AiTwotoneDelete } from 'react-icons/ai';
+import { FaHeart } from 'react-icons/fa';
+
+import styles from './styles.css';
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx:13]: Move relative imports after absolute imports

File: app/views/CartPage/index.tsx [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+        cartItems (createdBy: $email, page: $page, pageSize: $pageSize){
+            results {
+                id
+                totalPrice
+                book {
+                id
+                title
+                image {
+                    url
+                }
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx:26]: Please fetch id for image

File: app/views/CartPage/index.tsx [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+import { CartListQuery, CartListQueryVariables, DeleteCartItemsMutation, DeleteCartItemsMutationVariables } from '#generated/types';
+
+const CART_LIST = gql`
+    query CartList ($email: ID!, $page: Int!, $pageSize: Int!) {
+        cartItems (createdBy: $email, page: $page, pageSize: $pageSize){
+            results {
+                id
+                totalPrice
+                book {
+                id
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: The indentation is not correct

File: app/views/CartPage/index.tsx [hunk 16]
```diff
@@ -0,0 +1,201 @@
...
+interface Author {
+    id: number;
+    name: string;
+}
+
+interface Image {
+    url: string;
+}
+
+interface Book {
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx:61]:
```suggestion
interface CartContentProps {
```

File: app/views/CartPage/index.tsx [hunk 17]
```diff
@@ -0,0 +1,201 @@
...
+    id: number;
+    price: number;
+    authors: Author[];
+    quantity: number;
+    image: Image;
+    title: string;
+    cartId: string;
+    removeBookFromCartList: (id: string) => void;
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx:69]:
```suggestion
    onBookRemoveFromCartList: (id: string) => void;
```

File: app/views/CartPage/index.tsx [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+                    <div className={styles.headerDescription}>
+                        <TextOutput
+                            label="author"
+                            value={authorsDisplay}
+                        />
+                        <TextOutput
+                            label="Price (NPR)"
+                            valueType="number"
+                            value={price}
+                        />
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx:109]: Use number output

File: app/views/CartPage/index.tsx [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+                        <div className={styles.quantity}>
+                            <TextOutput
+                                label="Quantity"
+                                valueType="number"
+                            />
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx:114]: Use number output

File: app/views/CartPage/index.tsx [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+                            <NumberInput
+                                name="quantity"
+                                value={quantity}
+                                onChange={undefined}
+                            />
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx:119]: Why use NumberInput here without onChange

File: app/views/CartPage/index.tsx [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+                        icons={<FaHeart />}
+                        autoFocus
+                    >
+                        Add to wish list
+                    </Button>
+                    <Button
+                        name={undefined}
+                        variant="secondary"
+                        icons={<AiTwotoneDelete />}
+                        autoFocus
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: Why add autoFocus to two buttons?

File: app/views/CartPage/index.tsx [hunk 22]
```diff
@@ -0,0 +1,201 @@
...
+                                value={quantity}
+                                onChange={undefined}
+                            />
+                        </div>
+                    </div>
+                </Container>
+                <div className={styles.wishListButton}>
+                    <Button
+                        name={undefined}
+                        onClick={undefined}
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx:126]: Let's handle this button or disable this button and add a fixme/todo

File: app/views/CartPage/index.tsx [hunk 23]
```diff
@@ -0,0 +1,201 @@
...
+    };
+    return (
+        <>
+            <div className={styles.wishList}>
+                {!loading && data?.cartItems?.results
+                    && (
+                        <>
+                            {
+                                data.cartItems.results.map((b: any) => (
+                                    <CartContent
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: Please use ListView.
ListView handles all of these cases like loading, empty, etc.

File: app/views/CartPage/index.tsx [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+        refetch();
+    };
+    return (
+        <>
+            <div className={styles.wishList}>
+                {!loading && data?.cartItems?.results
+                    && (
+                        <>
+                            {
+                                data.cartItems.results.map((b: any) => (
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx:173]: Let's not use `any` type

File: app/views/CartPage/index.tsx [hunk 25]
```diff
@@ -0,0 +1,201 @@
...
+    >(DELETE_CART_ITEMS);
+
+    const { data, refetch, loading } = useQuery<
+        CartListQuery,
+        CartListQueryVariables
+    >(CART_LIST, {
+        variables: { page, pageSize, email },
+    });
+
+    const removeCartItem = (id: string) => {
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx:162]: Please use useCallback

File: app/views/CartPage/index.tsx [hunk 26]
```diff
@@ -0,0 +1,201 @@
...
+                        </>
+                    )}
+            </div>
+            <Button
+                name={undefined}
+                variant="secondary"
+                autoFocus
+            >
+                Order
+            </Button>
```
Comment (tnagorra) [inline app/views/CartPage/index.tsx]: Let's handle this or disable this button and add a FIXME/TODO comment

Also, why add "autoFocus" on button?

File: app/views/CartPage/styles.css [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+                    .quantity {
+                        display: flex;
+                    }
+                }
+            }
+
+            .wish-list-button {
+                button {
+                    margin-top: 40px;
+                }
```
Comment (tnagorra) [inline app/views/CartPage/styles.css:40]: let's not use asymmetric spacing to fix styling issues

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/views/CartPage/index.tsx (hunk 14 above)
Comment (barshathakuri) [reply app/views/CartPage/index.tsx:26]: There is no id for image in API

**Pull request closed without merging**